File: packages/faro-android-sdk/src/main/kotlin/com/edkimmel/faro/persistence/PreInitBuffer.kt
package com.edkimmel.faro.persistence

import com.edkimmel.faro.internal.Clock
import kotlinx.serialization.Serializable
import kotlinx.serialization.encodeToString
import kotlinx.serialization.json.Json
import java.io.File
import java.util.UUID

/**
 * Configuration for the pre-initialization buffer.
 */
data class PreInitBufferConfig(
    /** Maximum disk usage for pre-init buffered signals. */
    val maxDiskUsageBytes: Long = 1L * 1024 * 1024,
    /** Maximum age of buffered signals before they are discarded (default: 72 hours). */
    val maxFileAgeMs: Long = 72L * 60 * 60 * 1000,
    /** Directory name within the Faro cache directory. */
    val dirName: String = "faro_pre_init"
)

/**
 * A lightweight, static disk buffer for capturing signals before the SDK is initialized.
 *
 * Signals are written as individual JSON files to a known directory. When the SDK
 * initializes, it reads these files, replays them through the normal pipeline, and
 * deletes them.
 *
 * This is intentionally simple — no session, user, or device meta is attached at
 * write time. Meta is attached during replay when the full SDK context is available.
 */
internal class PreInitBuffer private constructor() {

    companion object {
        @Volatile
        private var instance: PreInitBuffer? = null

        fun getInstance(): PreInitBuffer {
            return instance ?: synchronized(this) {
                instance ?: PreInitBuffer().also { instance = it }
            }
        }
    }

    private val json = Json {
        ignoreUnknownKeys = true
        encodeDefaults = false
        explicitNulls = false
    }

    private var config = PreInitBufferConfig()
    private var bufferDir: File? = null

    fun configure(cacheDir: File, config: PreInitBufferConfig = PreInitBufferConfig()) {
        synchronized(this) {
            this.config = config
            this.bufferDir = File(File(cacheDir, "faro"), config.dirName).apply { mkdirs() }
        }
    }

    private fun ensureDir(): File? = bufferDir

    // ---- Write methods ----

    fun bufferLog(message: String, level: String, context: Map<String, String>? = null) {
        val entry = PreInitEntry(
            signalType = "log",
            timestamp = Clock.nowTimestamp(),
            log = PreInitLogData(message = message, level = level, context = context)
        )
        writeEntry(entry)
    }

    fun bufferError(
        type: String,
        value: String,
        stacktrace: PreInitStacktrace? = null,
        context: Map<String, String>? = null
    ) {
        val entry = PreInitEntry(
            signalType = "exception",
            timestamp = Clock.nowTimestamp(),
            exception = PreInitExceptionData(
                type = type, value = value, stacktrace = stacktrace, context = context
            )
        )
        writeEntry(entry)
    }

    fun bufferMeasurement(type: String, values: Map<String, Double>, context: Map<String, String>? = null) {
        val entry = PreInitEntry(
            signalType = "measurement",
            timestamp = Clock.nowTimestamp(),
            measurement = PreInitMeasurementData(type = type, values = values, context = context)
        )
        writeEntry(entry)
    }

    fun bufferEvent(name: String, attributes: Map<String, String>? = null, domain: String? = null) {
        val entry = PreInitEntry(
            signalType = "event",
            timestamp = Clock.nowTimestamp(),
            event = PreInitEventData(name = name, domain = domain, attributes = attributes)
        )
        writeEntry(entry)
    }

    // ---- Read and replay ----

    fun readPendingEntries(): List<PreInitEntry> {
        val dir = ensureDir() ?: return emptyList()
        enforceStorageLimits()

        return dir.listFiles()
            ?.filter { it.isFile && it.extension == "json" }
            ?.sortedBy { it.name }
            ?.mapNotNull { file ->
                try {
                    json.decodeFromString<PreInitEntry>(file.readText())
                } catch (_: Exception) {
                    file.delete()
                    null
                }
            } ?: emptyList()
    }

    fun clear() {
        val dir = ensureDir() ?: return
        dir.listFiles()?.forEach { it.delete() }
    }

    fun isEmpty(): Boolean {
        val dir = ensureDir() ?: return true
        return (dir.listFiles()?.size ?: 0) == 0
    }

    // ---- Private ----

    private fun writeEntry(entry: PreInitEntry) {
        val dir = ensureDir() ?: return
        try {
            enforceStorageLimits()
            val fileName = "${System.currentTimeMillis()}_${UUID.randomUUID()}.json"
            val file = File(dir, fileName)
            file.writeText(json.encodeToString(entry))
        } catch (_: Exception) {
            // Best effort — no logger available pre-init
        }
    }

    private fun enforceStorageLimits() {
        val dir = ensureDir() ?: return

        try {
            val now = System.currentTimeMillis()
            val allFiles = dir.listFiles()?.toList() ?: return

            // Delete expired files
            for (file in allFiles) {
                if (now - file.lastModified() > config.maxFileAgeMs) {
                    file.delete()
                }
            }

            // Enforce size limit
            val remaining = dir.listFiles()?.sortedBy { it.name } ?: return
            var totalSize = remaining.sumOf { it.length() }

            if (totalSize > config.maxDiskUsageBytes) {
                for (file in remaining) {
                    if (totalSize <= config.maxDiskUsageBytes) break
                    val fileSize = file.length()
                    file.delete()
                    totalSize -= fileSize
                }
            }
        } catch (_: Exception) {
            // Best effort
        }
    }
}

// ---- Pre-init entry models ----

@Serializable
internal data class PreInitEntry(
    val signalType: String,
    val timestamp: String,
    val log: PreInitLogData? = null,
    val exception: PreInitExceptionData? = null,
    val measurement: PreInitMeasurementData? = null,
    val event: PreInitEventData? = null
)

@Serializable
internal data class PreInitLogData(
    val message: String,
    val level: String,
    val context: Map<String, String>? = null
)

@Serializable
internal data class PreInitExceptionData(
    val type: String,
    val value: String,
    val stacktrace: PreInitStacktrace? = null,
    val context: Map<String, String>? = null
)

@Serializable
internal data class PreInitStacktrace(
    val frames: List<PreInitStackFrame>
)

@Serializable
internal data class PreInitStackFrame(
    val filename: String,
    val function: String,
    val colno: Int? = null,
    val lineno: Int? = null
)

@Serializable
internal data class PreInitMeasurementData(
    val type: String,
    val values: Map<String, Double>,
    val context: Map<String, String>? = null
)

@Serializable
internal data class PreInitEventData(
    val name: String,
    val domain: String? = null,
    val attributes: Map<String, String>? = null
)

File: packages/faro-ios-sdk/Sources/FaroSDK/Persistence/PreInitBuffer.swift
import Foundation

/// Configuration for the pre-initialization buffer.
public struct PreInitBufferConfig {
    /// Maximum disk usage for pre-init buffered signals.
    public let maxDiskUsageBytes: Int64

    /// Maximum age of buffered signals before they are discarded.
    public let maxFileAgeSeconds: TimeInterval

    /// Directory name within the Faro cache directory.
    public let dirName: String

    public init(
        maxDiskUsageBytes: Int64 = 1 * 1024 * 1024,
        maxFileAgeSeconds: TimeInterval = 72 * 60 * 60,
        dirName: String = "faro_pre_init"
    ) {
        self.maxDiskUsageBytes = maxDiskUsageBytes
        self.maxFileAgeSeconds = maxFileAgeSeconds
        self.dirName = dirName
    }
}

/// A lightweight, static disk buffer for capturing signals before the SDK is initialized.
///
/// Signals are written as individual JSON files to a known directory. When the SDK
/// initializes, it reads these files, replays them through the normal pipeline, and
/// deletes them.
///
/// This is intentionally simple — no session, user, or device meta is attached at
/// write time. Meta is attached during replay when the full SDK context is available.
internal final class PreInitBuffer {
    static let shared = PreInitBuffer()

    private let fileManager = FileManager.default
    private let writeQueue = DispatchQueue(label: "com.edkimmel.faro.preinitbuffer")
    private var config = PreInitBufferConfig()
    private var bufferDir: URL

    private init() {
        let cacheDir = FileManager.default.urls(for: .cachesDirectory, in: .userDomainMask).first!
        bufferDir = cacheDir.appendingPathComponent("faro").appendingPathComponent(config.dirName)
        try? fileManager.createDirectory(at: bufferDir, withIntermediateDirectories: true)
    }

    func configure(_ config: PreInitBufferConfig) {
        writeQueue.sync {
            self.config = config
            let cacheDir = FileManager.default.urls(for: .cachesDirectory, in: .userDomainMask).first!
            self.bufferDir = cacheDir.appendingPathComponent("faro").appendingPathComponent(config.dirName)
            try? self.fileManager.createDirectory(at: self.bufferDir, withIntermediateDirectories: true)
        }
    }

    // MARK: - Write methods

    func bufferLog(message: String, level: LogLevel, context: [String: String]? = nil) {
        let entry = PreInitEntry(
            signalType: "log",
            timestamp: Clock.nowTimestamp(),
            log: PreInitLogData(message: message, level: level.rawValue, context: context)
        )
        writeEntry(entry)
    }

    func bufferError(type: String, value: String, stacktrace: Stacktrace? = nil, context: [String: String]? = nil) {
        let entry = PreInitEntry(
            signalType: "exception",
            timestamp: Clock.nowTimestamp(),
            exception: PreInitExceptionData(type: type, value: value, stacktrace: stacktrace, context: context)
        )
        writeEntry(entry)
    }

    func bufferMeasurement(type: String, values: [String: Double], context: [String: String]? = nil) {
        let entry = PreInitEntry(
            signalType: "measurement",
            timestamp: Clock.nowTimestamp(),
            measurement: PreInitMeasurementData(type: type, values: values, context: context)
        )
        writeEntry(entry)
    }

    func bufferEvent(name: String, attributes: [String: String]? = nil, domain: String? = nil) {
        let entry = PreInitEntry(
            signalType: "event",
            timestamp: Clock.nowTimestamp(),
            event: PreInitEventData(name: name, domain: domain, attributes: attributes)
        )
        writeEntry(entry)
    }

    // MARK: - Read and replay

    func readPendingEntries() -> [PreInitEntry] {
        enforceStorageLimits()

        guard let fileURLs = try? fileManager.contentsOfDirectory(
            at: bufferDir,
            includingPropertiesForKeys: nil,
            options: .skipsHiddenFiles
        ) else {
            return []
        }

        return fileURLs
            .filter { $0.pathExtension == "json" }
            .sorted { $0.lastPathComponent < $1.lastPathComponent }
            .compactMap { url in
                guard let data = try? Data(contentsOf: url) else { return nil }
                guard let entry = try? JSONDecoder().decode(PreInitEntry.self, from: data) else {
                    try? fileManager.removeItem(at: url)
                    return nil
                }
                return entry
            }
    }

    func clear() {
        guard let fileURLs = try? fileManager.contentsOfDirectory(
            at: bufferDir,
            includingPropertiesForKeys: nil,
            options: .skipsHiddenFiles
        ) else {
            return
        }
        for url in fileURLs {
            try? fileManager.removeItem(at: url)
        }
    }

    var isEmpty: Bool {
        guard let contents = try? fileManager.contentsOfDirectory(
            at: bufferDir,
            includingPropertiesForKeys: nil,
            options: .skipsHiddenFiles
        ) else {
            return true
        }
        return contents.isEmpty
    }

    // MARK: - Private

    private func writeEntry(_ entry: PreInitEntry) {
        writeQueue.async { [weak self] in
            guard let self = self else { return }
            self.enforceStorageLimits()

            let fileName = "\(Int(Date().timeIntervalSince1970 * 1000))_\(UUID().uuidString).json"
            let fileURL = self.bufferDir.appendingPathComponent(fileName)

            do {
                let data = try JSONEncoder().encode(entry)
                try data.write(to: fileURL, options: .atomic)
            } catch {
                // Best effort — no logger available pre-init
            }
        }
    }

    private func enforceStorageLimits() {
        guard let fileURLs = try? fileManager.contentsOfDirectory(
            at: bufferDir,
            includingPropertiesForKeys: [.fileSizeKey, .contentModificationDateKey],
            options: .skipsHiddenFiles
        ) else {
            return
        }

        let now = Date()

        // Delete expired files
        for url in fileURLs {
            if let attrs = try? fileManager.attributesOfItem(atPath: url.path),
               let modDate = attrs[.modificationDate] as? Date,
               now.timeIntervalSince(modDate) > config.maxFileAgeSeconds {
                try? fileManager.removeItem(at: url)
            }
        }

        // Enforce size limit
        guard let remaining = try? fileManager.contentsOfDirectory(
            at: bufferDir,
            includingPropertiesForKeys: nil,
            options: .skipsHiddenFiles
        ) else {
            return
        }

        let sorted = remaining.sorted { $0.lastPathComponent < $1.lastPathComponent }
        var totalSize: Int64 = 0
        for url in sorted {
            if let attrs = try? fileManager.attributesOfItem(atPath: url.path),
               let size = attrs[.size] as? Int64 {
                totalSize += size
            }
        }

        if totalSize > config.maxDiskUsageBytes {
            for url in sorted {
                guard totalSize > config.maxDiskUsageBytes else { break }
                if let attrs = try? fileManager.attributesOfItem(atPath: url.path),
                   let size = attrs[.size] as? Int64 {
                    try? fileManager.removeItem(at: url)
                    totalSize -= size
                }
            }
        }
    }
}

// MARK: - Pre-init entry models

internal struct PreInitEntry: Codable {
    let signalType: String
    let timestamp: String
    let log: PreInitLogData?
    let exception: PreInitExceptionData?
    let measurement: PreInitMeasurementData?
    let event: PreInitEventData?

    init(
        signalType: String,
        timestamp: String,
        log: PreInitLogData? = nil,
        exception: PreInitExceptionData? = nil,
        measurement: PreInitMeasurementData? = nil,
        event: PreInitEventData? = nil
    ) {
        self.signalType = signalType
        self.timestamp = timestamp
        self.log = log
        self.exception = exception
        self.measurement = measurement
        self.event = event
    }
}

internal struct PreInitLogData: Codable {
    let message: String
    let level: String
    let context: [String: String]?
}

internal struct PreInitExceptionData: Codable {
    let type: String
    let value: String
    let stacktrace: Stacktrace?
    let context: [String: String]?
}

internal struct PreInitMeasurementData: Codable {
    let type: String
    let values: [String: Double]
    let context: [String: String]?
}

internal struct PreInitEventData: Codable {
    let name: String
    let domain: String?
    let attributes: [String: String]?
}
